src/WCSimPrimaryGeneratorAction.cc

```diff
@@ -910,8 +913,116 @@ void WCSimPrimaryGeneratorAction::GeneratePrimaries(G4Event* anEvent)
       	*/
       }
 
+    } else if (useDataTableEvt) {
+    // Setup local variables to store the data table values
+    G4int index;
+    G4int pdgid;
+    G4double ene;
+    G4ThreeVector pos;
+    G4ThreeVector dir;
+    G4double t;
+
+    // Check if the input file is open
+    if (!inputFile.is_open()) {
+      G4cout << "Error: Input file is not open" << G4endl;
+      G4cout << "Set a vector file using the command /mygen/vecfile <FILENAME>"
+             << G4endl;
+      exit(-1);
+    }
+
+    // Flag for first event
+    G4bool firstParticle = true;
+
+    // Line position before the next particle
+    std::streampos lastLinePos;
+
+    // Counter for the number of particles per event
+    G4int nParticles = 0;
+
+    // Read the data table
+    std::string line;
+
+    // Skip any lines that start with a #
+    while (std::getline(inputFile, line)) {
+      if (line.empty() || line[0] == '#') {
+        continue;
+      }
+
+      // If we've reached the end of the file before beamOn events have been reached then we need to stop
+      if (inputFile.eof()) {
+        G4cout << "End of datatable file - run terminated..." << G4endl;
+        G4RunManager::GetRunManager()->AbortRun();
+      }
+
+      if(nParticles > MAX_N_VERTICES){
+        G4cout << "CANNOT DEAL WITH MORE THAN " << MAX_N_VERTICES << " VERTICES" << G4endl;
+        exit(-1);
+      }
+
+      // Buffer to convert between string and other variables
+      std::istringstream buffer(line);
+
+      // Load information into local variables
+      buffer >> index >> pdgid >> ene >> pos[0] >> pos[1] >> pos[2] >> dir[0] >>
+          dir[1] >> dir[2] >> t;
+
+      // We've reached the end of the event N that we're generating and have also read the first line of event
+      // N + 1. We need to rewind to the start of event N + 1 so that the first particle of N + 1 is not missed when
+      // this function is called again for the next event.
+      if (index == 0 && firstParticle == false) {
+        // Go back a line
+        inputFile.seekg(lastLinePos);
+        // Set the number of particles in this event
+        SetNvtxs(nParticles);
+        break;
+      }
+
+      // Buffer the position of the current line in the file
+      lastLinePos = inputFile.tellg();
+
+      // Set values to save to output
+      SetBeamEnergy(ene, nParticles);
+      SetBeamDir(dir, nParticles);
+      SetBeamPDG(pdgid, nParticles);
+      SetVtxs(nParticles, pos);
+
+      nParticles++;
+
```

Comment:
```suggestion
SetNvtxs(nParticles);
```
I think this should be here, so if the last line of the file is read as an event you want to generate, `nvtxs` is set correctly

Am I getting the logic right?

Reply:
I see what you mean, though would this not call `SetNvtxs(nParticles)` every time the loop is run? So it would be called for every particle in the event.

I think it might actually make better sense outside of the while loop after all the particles have been looped over. This would then catch all of the events as long as `MAX_N_VERTICES` isn't reached.

Reply:
Yeah that's a better fix thanks